conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/TezosGovernanceOperations.scala

```diff
@@ -0,0 +1,203 @@
+    *                         `proposalsBlocks` to be available, plus the level immediately before this batch.
+    *                          ideally this would be a `Level => List[Voting.BakerRolls]` (i.e. a total function)
+    * @param ec needed to compose concurrent operations
+    * @return aggregated data
+    */
+  def aggregateData(db: Database)(
+      proposalsBlocks: Map[Block, ProtocolId],
+      proposalsBallots: Map[Block, List[Voting.Ballot]],
+      levelsRolls: Map[Int, List[Voting.BakerRolls]]
+  )(
+      implicit ec: ExecutionContext
+  ): Future[List[GovernanceAggregate]] = {
+
+    //local functions, should simplify reading the calling code
+    def countBallotsPerLevel(block: Block) =
+      TezosDatabaseOperations.getBallotOperationsForLevel(block.data.header.level)
+
+    // as stated, this is a runtime failure if the block contains the wrong metadata
+    def countBallotsPerCycle(block: Block) = block.data.metadata match {
+      case BlockHeaderMetadata(_, _, _, _, _, level) =>
+        TezosDatabaseOperations.getBallotOperationsForCycle(level.cycle)
+    }
+
+    // as stated, this is a runtime failure if the block contains the wrong metadata
+    def proposalOperationsHashes(block: Block) = block.data.metadata match {
+      case BlockHeaderMetadata(_, _, _, _, _, level) =>
+        TezosDatabaseOperations.getProposalOperationHashesByCycle(level.cycle)
+    }
+
+    logger.info("Searching for governance data in voting period...")
+
+    //main algorithm
+    val ballotCountsResult =
+      Future
+        .traverse(proposalsBlocks.keys) { block =>
+          val cycleQuery = db.run(countBallotsPerCycle(block))
+          val levelQuery = db.run(countBallotsPerLevel(block))
+          (cycleQuery, levelQuery).mapN(
+            (cycleCounts, levelCounts) => block -> ((BallotsPerCycle(cycleCounts), BallotsPerLevel(levelCounts)))
+          )
+        }
+        .map(_.toMap)
+
+    val proposalCountsResult = Future.traverse(proposalsBlocks.keys) { block =>
+      db.run(proposalOperationsHashes(block).map(block -> _))
+    }
+
+    for {
+      countsMap <- ballotCountsResult
+      proposalCounts <- proposalCountsResult
+    } yield
+      fillAggregates(
+        proposalsBlocks,
+        levelsRolls,
+        proposalsBallots,
+        countsMap,
+        proposalCounts.toMap
+      )
+  }
+
+  /* Having all data ready, we can process per block,
+   * extract the numbers returning a collector object.
+   */
+  private def fillAggregates(
+      proposalsBlocks: Map[Block, ProtocolId],
+      listingsPerLevel: Map[Int, List[Voting.BakerRolls]],
+      ballots: Map[Block, List[Voting.Ballot]],
+      ballotCounts: Map[Block, (BallotsPerCycle, BallotsPerLevel)],
+      proposalProtocolCounts: Map[Block, Map[ProtocolId, Int]] //comes from individual operations on the block
+  ): List[GovernanceAggregate] =
+    proposalsBlocks.toList.flatMap {
```

Comment:
I think we have a bug here because we will be missing blocks from the proposal period.

Reply:
That's the subject of the following PR, which is already waiting